### ISSUE TITLE
Cannot GET /android callback issue 

### ISSUE THREAD

**Issue by f0ster** [auth0-samples/auth0-android-sample#65]
I am still getting this issue and I have followed and read all of the documentation references @lbalmaceda referenced in #49 et. al.

I am using 00-login successfully with facebook with the demo scheme, and the demo:// callback url in my approved callbacks.

If I take this working example, and change the scheme to https in the manifest and in the lock builder, then I have the option to choose Browser or Hosted Login Page.  I choose browser, and the redirect fails 
with the same `Cannot GET /android/com.xyx.abc/callback` error.  I have the approved callback added in the admin as expect, e.g, `https://mydomain.auth0.com/android/com.company.project/callback`, 

The login via the hosted page works, and I would prefer the user didn't have an option unless necessary

Please advise,

TLDR

1. Use 00-Login example
2. Replace references to `demo` scheme with `https`
3. add approved callback in admin, `https://mydomain.auth0.com/android/com.company.project/callback`
4. choose browser when prompted in android during auth

**Comment by lbalmaceda** [comment]
@f0ster You said you were prompted to choose either `Browser` or `Hosted Login Page`. I guess the latter is your application name, am I correct? If so, that's normal since both browser applications and your application can handle an URL like `https://mydomain.auth0.com/android/com.company.project/callback`. The problem is that after you've logged in, what the dialog really asks is "who should handle the token response" and the browser app has no idea how to deal with Auth0 tokens. **The only flow that will succeed is choosing your application.** To avoid this dialog you have 2 options:

- Require minimum API 23 and [use android app links](https://github.com/auth0/Auth0.Android#a-note-about-app-deep-linking), where you provide a domain verification file that tells Android you own "mydomain.auth0.com" and that _that_ callback should be handled only in _your_ android application OR
- Use a different scheme that no other app uses, i.e. `demo` as in this samples.

First step is safer but limits you to devices using API 23 and up. If your app is already targeting that SDK, then I'd say go for 1. Otherwise, 2 is your choice.

**Comment by f0ster** [comment]
Thank you for the clarification @lbalmaceda :fire: :100:

**Comment by CarsonRedeye** [comment]
I'm not even prompted to choose between browser or app, it defaults to browser. Therefore changing the scheme to something other than https is a must

**Comment by lbalmaceda** [comment]
@CarsonRedeye you must have already chosen a default handler app for that type of intent. You can remove the defaults in the settings. You normally do this in `Settings > Applications > Manage applications` find the application and click Clear defaults.

**Comment by kaleempeeroo** [comment]
Hi. I tried the steps mentioned with 
[https://github.com/auth0-samples/auth0-android-sample/issues/49](#49)
but still cannot get callback from the android app. Plz advise.

**Comment by kaleempeeroo** [comment]
Hi. I tried the steps mentioned:
#65
#49
but still cannot get callback from the android app. Plz advise.

Signup Activity:
https://github.com/kaleempeeroo/wheresmydosh/blob/ClientLogin/client/app/src/main/java/com/helpathand/helpathand/SignupActivity.kt

Manifest:
https://github.com/kaleempeeroo/wheresmydosh/blob/ClientLogin/client/app/src/main/AndroidManifest.xml

App build.gradle:
https://github.com/kaleempeeroo/wheresmydosh/blob/ClientLogin/client/app/build.gradle#L43

Allowed URL:
com.helpathand.helpathand://dev-wheresmydosh.auth0.com/android/com.helpathand.helpathand

**Comment by lbalmaceda** [comment]
@kaleempeeroo I tried your sample and the error I got into is pretty self-explanatory:

![image](https://user-images.githubusercontent.com/3900123/62470476-e48a4d80-b770-11e9-858c-060160415233.png)


The callback URL you need to whitelist in the Allowed Callback URL section of your application should be changed to the following one, as explained in https://github.com/auth0/Lock.Android/#usage.

```
com.helpathand.helpathand://dev-wheresmydosh.auth0.com/android/com.helpathand.helpathand/callback
```

You can check that if you enable logging on the `Auth0` instance (remember to remove it on prod):

```java
auth0_account.isLoggingEnabled = true
```

then on the logcat
```
com.helpathand.helpathand V/CallbackHelper: The Callback URI is: com.helpathand.helpathand://dev-wheresmydosh.auth0.com/android/com.helpathand.helpathand/callback
```

**Comment by kaleempeeroo** [comment]
Thanks for the clarifications. Did receive the same error in logcat.

So, the Allowed Callback URL should be changed to : 
https://dev-wheresmydosh.auth0.com/android/com.helpathand.helpathand/callback

I did try it a number of times without success. Should the scheme "https" and url be declared only in the Application Allowed Callback URLs because in #65, they had to do it in 3 different places? I tried it but that didnt work either.

**Comment by lbalmaceda** [comment]
@kaleempeeroo If you are using `Lock` then configure it with the `withScheme` as you have been doing so far. If you only want to use the `WebAuthProvider` class without using Lock at all, then the same applies.. when you configure the WebAuthProvider build pass `withScheme`.

The scheme for your android app must be declared both in the manifestPlaceholders of the build.gradle file, as well as in the withScheme method call as [explained here](https://github.com/auth0/Auth0.Android#a-note-about-app-deep-linking). I saw you are using a custom scheme, different than https. It should work anyway as long as you change both places in your app as mentioned above and update the Allowed Callback URL with the one the logcat says, which will match the scheme you requested. 

What error are you running into?

**Comment by kaleempeeroo** [comment]
I think I would like to use Lock.
Tried 'https' but got the same not allowed url error so tried custom scheme with same result.
Changed both places in app as you mentioned above and updated the Allowed Callback URL, still not working. Below is logcat output. Has it got to do with conflicting SDKs and libraries am using?
Using:
Compile SDK Version 28 and running on Pixel 2 emulator with API 29. Tried Compile SDK Version 29 as well. Lock version 2.15.0

`08/06 09:56:09: Launching app
$ adb install-multiple -r -t C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\resources\instant-run\debug\resources-debug.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\dep\dependencies.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_6.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_1.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_0.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_5.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_8.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_7.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_9.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_4.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_3.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\split-apk\debug\slices\slice_2.apk C:\Users\server\Dropbox\Shah Informatics\projects\CC App\wheresmydosh\client\app\build\intermediates\instant-run-apk\debug\app-debug.apk 
Split APKs installed in 22 s 176 ms
$ adb shell am start -n "com.helpathand.helpathand/com.helpathand.helpathand.SignupActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Client not ready yet..Waiting for process to come online
Connected to process 3833 on device emulator-5554
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
W/hand.helpathan: JIT profile information will not be recorded: profile file does not exits.
I/chatty: uid=10150(com.helpathand.helpathand) identical 10 lines
W/hand.helpathan: JIT profile information will not be recorded: profile file does not exits.
I/InstantRun: starting instant run server: is main process
V/Builder: Lock instance created
    Using Telemetry Lock.Android (2.15.0) and Library 1.18.0
W/hand.helpathan: Accessing hidden method Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z (greylist, reflection, allowed)
    Accessing hidden method Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V (greylist, reflection, allowed)
D/: HostConnection::get() New Host Connection established 0xde3582d0, tid 4094
    HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_dma_v1 ANDROID_EMU_host_composition_v1 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_gles_max_version_2 
D/OpenGLRenderer: Swap behavior 1
D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 0 0
D/EGL_emulation: eglCreateContext: 0xde31a120: maj 2 min 0 rcv 2
D/EGL_emulation: eglMakeCurrent: 0xde31a120: ver 2 0 (tinfo 0xde30f3b0)
W/Gralloc3: mapper 3.x is not supported
D/: createUnique: call
    HostConnection::get() New Host Connection established 0xde358410, tid 4094
    HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_dma_v1 ANDROID_EMU_host_composition_v1 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_gles_max_version_2 
D/OpenGLRenderer: Setting buffer count to 3, min_undequeued 1, extraBuffers 0
D/EGL_emulation: eglMakeCurrent: 0xde31a120: ver 2 0 (tinfo 0xde30f3b0)
D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 1 0
I/OpenGLRenderer: Davey! duration=772ms; Flags=1, IntendedVsync=515948484697, Vsync=516031818027, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=516046843477, AnimationStart=516046879192, PerformTraversalsStart=516046882471, DrawStart=516292563814, SyncQueued=516293719058, SyncStart=516318708988, IssueDrawCommandsStart=516318955339, SwapBuffers=516573512810, FrameCompleted=516745499342, DequeueBufferDuration=42000, QueueBufferDuration=181000, 
W/ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@46d43bc
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
D/OpenGLRenderer: Setting buffer count to 3, min_undequeued 1, extraBuffers 0
D/EGL_emulation: eglMakeCurrent: 0xde31a120: ver 2 0 (tinfo 0xde30f3b0)
W/hand.helpathan: Accessing hidden method Lcom/android/org/conscrypt/ConscryptFileDescriptorSocket;->setUseSessionTickets(Z)V (greylist,core-platform-api, reflection, allowed)
    Accessing hidden method Lcom/android/org/conscrypt/ConscryptFileDescriptorSocket;->setHostname(Ljava/lang/String;)V (greylist,core-platform-api, reflection, allowed)
W/hand.helpathan: Accessing hidden method Lcom/android/org/conscrypt/OpenSSLSocketImpl;->setAlpnProtocols([B)V (greylist,core-platform-api, reflection, allowed)
W/hand.helpathan: Accessing hidden method Lcom/android/org/conscrypt/OpenSSLSocketImpl;->getAlpnSelectedProtocol()[B (greylist,core-platform-api, reflection, allowed)
I/ApplicationFetcher: Application received!
W/Configuration: You've chosen 'null' as your default database name, but it wasn't found in your Auth0 connections configuration.
V/FormLayout: Showing the LogIn/SignUp tabs
V/SocialView: New instance created. Using small buttons: false
D/FormLayout: Mode changed to 0
I/chatty: uid=10150(com.helpathand.helpathand) identical 1 line
D/FormLayout: Mode changed to 0
V/ValidatedInputView: Setting up validation for field of type 2
V/ValidatedInputView: Setting up validation for field of type 1
V/ValidatedInputView: Setting up validation for field of type 0
V/ValidatedInputView: Setting up validation for field of type 1
V/ValidatedInputView: Setting up validation for field of type 5
V/EnterpriseConnectionMatcher: Creating a new instance to match 0 Enterprise Connections
V/ValidatedInputView: Setting up validation for field of type 9
V/ValidatedInputView: Setting up validation for field of type 5
V/ValidatedInputView: Setting up validation for field of type 5
V/ValidatedInputView: Setting up validation for field of type 1
V/LogInFormView: Multiple enterprise/database connections found.
V/ValidatedInputView: Input cleared and validation errors removed
I/chatty: uid=10150(com.helpathand.helpathand) identical 1 line
V/ValidatedInputView: Input cleared and validation errors removed
I/Choreographer: Skipped 79 frames!  The application may be doing too much work on its main thread.
I/OpenGLRenderer: Davey! duration=3197ms; Flags=0, IntendedVsync=518257736784, Vsync=519574403398, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=519575757319, AnimationStart=519575800008, PerformTraversalsStart=519575805184, DrawStart=520719458328, SyncQueued=520729255151, SyncStart=520772280890, IssueDrawCommandsStart=520772705745, SwapBuffers=521473385098, FrameCompleted=521498535353, DequeueBufferDuration=65000, QueueBufferDuration=326000, 
I/Choreographer: Skipped 41 frames!  The application may be doing too much work on its main thread.
I/OpenGLRenderer: Davey! duration=793ms; Flags=0, IntendedVsync=520813896550, Vsync=521497229856, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=521500658920, AnimationStart=521500690686, PerformTraversalsStart=521500693303, DrawStart=521501226966, SyncQueued=521501840555, SyncStart=521599981180, IssueDrawCommandsStart=521600066774, SwapBuffers=521678816623, FrameCompleted=521705112119, DequeueBufferDuration=95000, QueueBufferDuration=527000, 
D/FormLayout: Mode changed to 1
I/chatty: uid=10150(com.helpathand.helpathand) identical 1 line
D/FormLayout: Mode changed to 1
V/ValidatedInputView: Setting up validation for field of type 0
V/ValidatedInputView: Setting up validation for field of type 1
V/ValidatedInputView: Setting up validation for field of type 1
V/ValidatedInputView: Setting up validation for field of type 5
V/ValidatedInputView: Setting up validation for field of type 1
V/ValidatedInputView: Setting up validation for field of type 1
V/ValidatedInputView: Setting up validation for field of type 5
D/SignUpFormView: Adding 0 custom fields.
V/ValidatedInputView: Input cleared and validation errors removed
I/chatty: uid=10150(com.helpathand.helpathand) identical 1 line
V/ValidatedInputView: Input cleared and validation errors removed
I/Choreographer: Skipped 174 frames!  The application may be doing too much work on its main thread.
I/OpenGLRenderer: Davey! duration=3068ms; Flags=0, IntendedVsync=633309733545, Vsync=633409733541, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=633421431643, AnimationStart=633421461656, PerformTraversalsStart=633421649997, DrawStart=636326790240, SyncQueued=636333240082, SyncStart=636333408788, IssueDrawCommandsStart=636333555683, SwapBuffers=636352458476, FrameCompleted=636378026987, DequeueBufferDuration=103000, QueueBufferDuration=556000, 
I/OpenGLRenderer: Davey! duration=3024ms; Flags=0, IntendedVsync=633427407857, Vsync=636327407741, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=636337257378, AnimationStart=636337302016, PerformTraversalsStart=636340018316, DrawStart=636340411262, SyncQueued=636345702259, SyncStart=636379067423, IssueDrawCommandsStart=636379275428, SwapBuffers=636458434432, FrameCompleted=636485109725, DequeueBufferDuration=74000, QueueBufferDuration=509000, 
V/LockActivity: Looking for a provider to use /authorize with the connection google-oauth2
D/LockActivity: Couldn't find an specific provider, using the default: WebAuthProvider
V/CallbackHelper: The Callback URI is: com.helpathand.helpathand://dev-wheresmydosh.auth0.com/android/com.helpathand.helpathand/callback
V/OAuthManager: Using PKCE authentication flow
W/ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@d3155e2
V/CustomTabsController: Trying to bind the service
V/CustomTabsController: Bind request result: true
D/CustomTabsController: CustomTabs Service connected
D/CustomTabsController: Launching URI. Custom Tabs available: true
D/OpenGLRenderer: Setting buffer count to 3, min_undequeued 1, extraBuffers 0
D/EGL_emulation: eglMakeCurrent: 0xde31a120: ver 2 0 (tinfo 0xde30f3b0)
D/EGL_emulation: eglMakeCurrent: 0xde31a120: ver 2 0 (tinfo 0xde30f3b0)
`

**Comment by lbalmaceda** [comment]
@kaleempeeroo Please share the error you are running into, either a screenshot or snippet. I don't see the error in what you've shared above.

**Comment by kaleempeeroo** [comment]
![phone](https://user-images.githubusercontent.com/30236761/62553131-dd9e3080-b87f-11e9-9e46-e20fecf3aa1c.jpg)

**Comment by kaleempeeroo** [comment]
it's finally working mate! Lost so much time with this mess.
As you saw, it was showing no errors so got no clue what was really going on there.
What I did was just create a new Auth0 Application and everything just flows smoothly.
Thanks for your clarifications and help though. Appreciate it.
![phone](https://user-images.githubusercontent.com/30236761/62558610-7639ae00-b88a-11e9-93b7-d6e84b5f74b7.jpg)

**Comment by lbalmaceda** [comment]
@kaleempeeroo the error page you attached here https://github.com/auth0-samples/auth0-android-sample/issues/65#issuecomment-518719000 is actually saying what the error was. The callback URL created by this SDK (using domain, scheme, package name, etc) was not registered in the "Allowed callback URLs" section of your native application in the dashboard. If you have scrolled down the page and clicked the "show details" link the error would explicitly said what callback URL was attempted to be used (should have been the same as in the logcat output) but wasn't whitelisted.

I'm glad it worked out in the end. Cheers

**Comment by ohassine** [comment]
@lbalmaceda 
I am getting same error (Cannot GET /android/com.xyx.abc/callback) when I request login(only when setting prompt parameter to none) and when I try to logout
I am using Android App links with https schema and min API 23
Auth0 Android v1.19.1.

**Comment by lbalmaceda** [comment]
@ohassine If you're setting the `scheme` value to `https` on both the manifestPlaceholder for `auth0Scheme` and the `WebAuthProvider.withScheme(..)`, the configuration from the SDK side should be correct. 

Things you need to check on the Auth0 dashboard:
- Auth0 app is of type NATIVE
- Callback URL is correctly whitelisted
- Advanced settings -> Device settings. Android section is filled with the package name and the SHA256 of your certificate (production and debug keys, just in case)

You should be able to test with [this tool](https://developers.google.com/digital-asset-links/tools/generator) that the configuration from the server is fine.

**Comment by ohassine** [comment]
@lbalmaceda I've rechecked these things, but I am still getting same error on logout

**Comment by kaleempeeroo** [comment]
@ohassine:  try delete the Auth0 application and recreate it with default settings.  
Check if working. If not, then add the whitelisted URL

**Comment by lbalmaceda** [comment]
@ohassine The logout callback (and its declared intent-filter) is the same as the one used for login. If login works, logout should work as well. You can use the logcat and check the logs that the OS prints when you install the app for the first time. There you can tell if the AppLink configuration was successful or not. I'd also check the settings of the app (once installed) and see if it mentions that links from that domain will be opened automatically by the app or not.
